### ISSUE TITLE
flags of carbonicity not working as expected

### ISSUE THREAD

**Issue by sishijing** [argiopetech/base#61]
When I set flags. --startingCarbonicity, --priorCarbonicity and --sigmaCarbonicity on the command line, the singlePopMcmc does not run at all. But I can run singlePopMcmc when I set values for these parameters in the base9.yaml file.

**Comment by maxvonhippel** [comment]
I think I've found the issue, looks like things set in command line if you also use YAML and/or the prompts get overridden and can cause issues.  I'm working on it now.  I think (unless I am missing some additional and large issue) I should have a functional fix up pretty soon, but this appears to be a problem where there are many different possible programmatic styles for the fix, so Travis may want me to code it a different way for stylistic reasons.  Anyway, I'm working on it.  Thanks for pointing this out!

**Comment by maxvonhippel** [comment]
What's happening is that the code does not check to see if a value was already set in the command line args before setting it based on YAML.  SO when you combine YAML with command line args your command line args are superseded but the YAML args.  Since you haven't updated to the latest version, when it fails to find anything in the YAML, it crashes (see getOrDie in Settings.cpp).  If you update to the latest version, it will not die, but instead ask you to re-enter the missing value.  That said, this behavior is clearly redundant.

What I plan to do is figure out what the defaults are for each of the requested values and then use ternaries to only update values which have not yet been set.  For example, according to Cluster.hpp, carbonicity is set by default to 0.0.  So in Settings.cpp, I could do something like:

````c++
cluster.starting.carbonicity = cluster.starting.carbonicity == 0.0 ? getOrRequest <double>(startingNode, "carbonicity") : cluster.starting.carbonicity;
````

Instead of

````c++
cluster.starting.carbonicity = getOrRequest <double>(startingNode, "carbonicity");
````

Because I will have to find the default value for each of the values to be checked, implementing this will be a bit time intensive, so I'll probably wait to do it until a bit later this week.  If possible I hope to also improve the prompt for variable entree using node name (for example, "please enter your desired general.verbose.cluster.starting.carbonicity" is much more helpful than "please enter your desired carbonicity" as there are multiple carbonicities to be entered).  

Hope that makes sense!

**Comment by maxvonhippel** [comment]
Sure, happy to.  I'm taking care of logging in the same PR, so this one will be a twofer.  I'm about 70% done.  You and Travis should probably chat about what the best format would be for the logs appended to the output file.  For example, should they look like a TSV?  Should it just be:

> valueName: value

... or ...

> full node path: (type) value

... or something else?  Anyway, I think unless I get a definitive answer soon, I'll just go with whatever is the most immediately programmatically element, and then we can see what people think and improve it moving forward.  

Fixing the issue identified by Sishijing means editing the same functions and code that I need to edit in order to add the logging which is why I'm just doing both at once.

**Comment by maxvonhippel** [comment]
@tedvh should default carbonicity be 0.38 for all carbonicity values (`cluster.starting. carbonicity `, `cluster.priors.means. carbonicity `, and `cluster.priors.sigmas. carbonicity `)?  Also, does this mean I should change the default for `singlePopMcmc.stepSizes.carbonicity`?

**Comment by tedvh** [comment]
Hi Max -

good question.  The defaults should be

0.38 for cluster.starting.carbonicity and
0.38 for cluster.priors.means.carbonicity

yet

0.0 for cluster.priors.sigmas.carbonicity

I don't know about singlePopMcmc.stepSizes.carbonicity .  That is a tuning 
parameter and should probably be left wherever Elliot put it for now.

-Ted

**Comment by maxvonhippel** [comment]
Sounds good, thanks.

**Comment by maxvonhippel** [comment]
@tedvh for some reason the Node `singlePopMcmc.burnIter` is called `stage2IterMax` in the YAML file.  Every single other variable is named the same thing in the code and the YAML.  Can I go ahead and change it to use `burnIter` in the YAML?  (Doing so would make my code easier to write, and I think it's more logical to use the same thing in both places anyway.)

**Comment by maxvonhippel** [comment]
I take back my comment about logging, I decided the approach I'd taken was too complex so I redid it to just address the particular problem raised here and not try to tackle logging simultaneously.  I'll work on that next in another pr.

**Comment by tedvh** [comment]
Hi Max -

Good eye.  Sure, go ahead and make that variable name change.  We have changed 
the burnin techniques a few times an probably that didn't propagate.

Elliot can also chime in if he recognizes a problem related to this.

-Ted

**Comment by tedvh** [comment]
OK, sounds good.

**Comment by maxvonhippel** [comment]
I deleted my comment about burnin because I found about 4 other variables with name mis-matches.  I want to make the names be the same in the YAML, the code, and the command line args, with only a slight difference in that YAML is of form 

> path.to.node[key] = variable

... and command line args look like:

--pathToNodeKey=variable

... and c++ variables are a bit abstracted, along the lines of:

````c++
// sudo code
Node config = whatever;
Node whiteDwarf = some stuff to get whiteDwarf from config;
variable whiteDwarf.someSpecificMetric = getOrRequest(... );
````

Anyway, I plan to go through and document all of the name mismatches and then open a new issue with them.  Then you and Elliot can decide what names to use throughout and I will update it accordingly.
